### PR TITLE
add multiselect functionality to listbox

### PR DESCRIPTION
**Pull request by radium-v**
# Pull Request

## 📖 Description
Adds the `multiple` attribute to `<fast-listbox>`.

### 🎫 Issues

* #4190 

## 👩‍💻 Reviewer Notes

This PR supersedes the earlier #4974 PR.

## 📑 Test Plan

The `<fast-listbox>` element should accept the `multiple` boolean attribute, which will put the component into a mode where more than one option can be selected by a user.

_For the purposes of this implementation, "focus" is the `checked` state of the listbox's options. While interacting with a listbox, focus should remain on the listbox and shouldn't be placed on the options. [See this note for more information](https://www.w3.org/TR/wai-aria-practices-1.2/#h-note-16)._

* The `ArrowDown` key should move focus to the next selectable option.
  * Holding `Shift` while pressing `ArrowDown` should toggle focus for the next selectable option.
* The `ArrowUp` key should move focus to the previous selectable option
  * Holding `Shift` while pressing `ArrowUp` should toggle focus for the previous selectable option.
* The `Home` key should move focus to the first selectable option.
  * Holding `Shift` while pressing `Home` should toggle focus for all options starting from the currently focused option to the first selectable option.
* The `End` key should move focus to the last selectable option
  * Holding `Shift` while pressing `End` should toggle focus for all options starting from the currently focused option to the last selectable option.
* The ` ` (Space) key should select all currently focused options.
  * If any focused options are unselected, all focused options will be selected.
  * If all focused options are selected, all focused options will be unselected.

### General

- [ ] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [x] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## ⏭ Next Steps

This PR is part of a series, split from #4974 to unblock improvements made independently of multiple selection mode.

### CODE REVIEW

**Comment by radium-v** [comment]
This PR is blocked by #5528.

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/listbox/listbox.ts:656]: Should this map to [ariaMultiSelectable](https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaMultiSelectable)?

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/listbox/listbox.ts:284]: If `0` is impossible do we need to be setting it on line 251 or can we simplify that ternary?

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/listbox/listbox.ts:208]: Is this preferred to DOM.queueUpdate in this scenario (curious)?

File: packages/web-components/fast-foundation/src/listbox/listbox.template.ts
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/listbox/listbox.template.ts:16]: The IDL (listed elsewhere) is [`ariaMultiSelectable`](https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaMultiSelectable), it seems like we'll want to map to that.

File: packages/web-components/fast-foundation/src/listbox/listbox.element.ts
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/listbox/listbox.element.ts:86]: This is nullable and set as such in combobox and select; should we apply it consistently?

**Review by chrisdholt — APPROVED** [review]

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/listbox/listbox.ts:208]: If you prefer, you could do `DOM.queueUpdate` here.

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/listbox/listbox.ts:480]: Interesting pattern here. I had considered an API for `Observable` like this:

```ts
Observable.stop(obj);
Observable.start(obj);

const notifier = Observable.getNotifier(obj);
notifier.isEnabled
```

Would that have been helpful in this case?

**Review by EisenbergEffect — APPROVED** [review]
I pulled this down and ran the Storybook. Very nice to see this in action. I left a couple of comments/questions. Code looks very clean. Definitely get more reviews from the components team as I skipped over a lot of the interaction/accessibility related details in my review and just looked for things that stood out from my perspective.

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (radium-v) [reply packages/web-components/fast-foundation/src/listbox/listbox.ts:208]: I wasn't able to get it to work with `DOM.queueUpdate`. The timing needs to guarantee that `focus()` and `scrollIntoView()` happen on different frames, and since this function is called by `openChanged`, the `updateQueue` ends up grouping them into the same frame.

The scenario to test it is with the "Select with very long list" storybook example:

1. Press `Tab` to focus the select
2. Press `End` to select the last item in the list
3. Press `Space` to open the list

The listbox should be scrolled to the bottom and "Wyoming" should be focused.

Alternatively, waiting for `DOM.nextUpdate` works here:
```ts
optionToFocus.focus();
DOM.nextUpdate().then(() => {
    optionToFocus.scrollIntoView({ block: "nearest" });
});
```

Comment (chrisdholt) [reply packages/web-components/fast-foundation/src/listbox/listbox.ts:208]: Good feedback here. Could we just add a comment for future proofing essentially saying this portion:

> The timing needs to guarantee that focus() and scrollIntoView() happen on different frames, and since this function is called by openChanged, the DOM.updateQueue ends up grouping them into the same frame, which is why we're using RAF

File: packages/web-components/fast-foundation/src/listbox/listbox.ts
Comment (radium-v) [reply packages/web-components/fast-foundation/src/listbox/listbox.ts]: Maybe something where you can perform a value without the notifier being called? There'd also have to be some way to configure if a notifier can be bypassed, and maybe by who/what.

```ts
const notifier = Observable.getNotifier(o);

notifier.bypass?.(this, "selected", () => {
  o.selected = filteredNext.includes(o);
});
```

**Pull request merged**
